Initial version of key exchange API

Brian-Heckel opened the pull request:
Includes an unfinished anshel anshel goldfield implementation and might need to refactor FiniteFieldDH to be implemented from CommutativeKeyExchange instead.

<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [ ] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

Brian-Heckel committed:
Initial Commit for the Key Exchange API

Includes an unfinished anshel anshel goldfield implementation
and might need to refactor FiniteFieldDH to be implemented from
CommutativeKeyExchange instead.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/anshel_anshel_goldfeld.py:
```diff
@@ -0,0 +1,50 @@
+        - B is a list of elements in the group G
+        """
+        self._group = G
+        # TODO: check that the group is
```
Is non-abelian required for correctness or security? If only for security, mention it in the docstring somewhere but no need to check at runtime.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/anshel_anshel_goldfeld.py:
```diff
@@ -0,0 +1,50 @@
+import random
+from sage.crypto.public_key.key_exchange import KeyExchange
+
+class AnshelAnshelGoldfeld(KeyExchange):
```
Can you split this file into another branch? I want to merge the rest of this before I'll have time to read up on this scheme so I can review it.

vincentmacri replied:
Also when you add the docstring, it should include an academic reference (like a textbook) to where someone can read about this scheme.

Use Sage's citing facilities. You might have to add an entry to the massive bibliography file. See my draft PR in this repo about function fields for how to do that if you don't know.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/anshel_anshel_goldfeld.py:
```diff
@@ -0,0 +1,50 @@
+        return pub_keys
+
+    def alice_compute_shared_secret(self, alice_sk, bob_pk):
+        beta = 
```
Lint error

vincentmacri commented on src/sage/crypto/public_key/key_exchange/commutative_key_exchange.py:
```diff
@@ -0,0 +1,35 @@
+from sage.crypto.public_key.key_exchange import KeyExchange
+from abc import ABC, abstractmethod
```
As far as I can tell this would be the first use of `ABC` or `abstractmethod` in Sage. Sage has its own `abstract_method`. I see no reason to use that here, and am more than happy to use the standard Python `abc` library. My best guess is the custom one is to better support the category stuff, but we don't need that here.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/commutative_key_exchange.py:
```diff
@@ -0,0 +1,35 @@
+from sage.crypto.public_key.key_exchange import KeyExchange
+from abc import ABC, abstractmethod
+
+class CommutativeKeyExchange(KeyExchange, ABC):
```
I think you could put this class in `key_exchange.py`. Also, since `KeyExchange` inherits from `ABC` isn't it redundant for `CommutativeKeyExchange` to also inherit from `ABC`?

Brian-Heckel replied:
- Agree with putting the class in key_exchange, would probably make importing eaiser
- I had it importing ABC because I wanted to have `secret_key` `public_key` and `compute_shared_secret` to be abstract methods, but I don't need to have it import ABC to have that. Will remove it

vincentmacri commented on src/sage/crypto/public_key/key_exchange/commutative_key_exchange.py:
```diff
@@ -0,0 +1,35 @@
+
+    def bob_compute_shared_secret(self, bob_sk, alice_pk):
+        return self.compute_shared_secret(bob_sk, alice_pk)
+
```
It would be nice to to exclude the `alice_foo` and `bob_foo` versions from tab completion for easier discoverability. You can do this with the dunder method `__dir__`. Make sure you write it in a way that will not exclude new methods implemented in subclasses though (i.e. just exclude the 6 methods that are implemented here).

Brian-Heckel replied:
Even if `__dir__` is changed so that `[secret_key, public_key, compute_shared_secret]` are the only methods shown, `__dir__` will be called on the `KeyExchange` subclass which includes all of the bob/alice methods. I think its unavoidable to have these methods shown

vincentmacri replied:
It's only in `CommutativeKeyExchange` where the alice/bob methods are uninteresting so that's fine. Anyway, not a huge deal. I might add `__dir__` myself after merging this if I remember to.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/finite_field_dh.py:
```diff
@@ -0,0 +1,30 @@
+from sage.rings.finite_rings.finite_field_constructor import GF
+from sage.misc.prandom import randint
+
+class FiniteFieldDH(KeyExchange):
```
This should inherit from `CommutativeKeyExchange`.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/finite_field_dh.py:
```diff
@@ -0,0 +1,30 @@
+        self.p = p
+        self.generator = self._field(generator)
+        # self.alice_key = randint(2, self.p - 2)
+        # self.bob_key = randint(2, self.p - 2)
```
```suggestion
```

vincentmacri commented on src/sage/crypto/public_key/key_exchange/key_exchange.py:
```diff
@@ -0,0 +1,56 @@
+        bob_sk = self.bob_secret_key()
+        bob_pk = self.bob_public_key(bob_sk)
+        shared_secret = self.alice_compute_shared_secret(alice_sk, bob_pk)
+        return (alice_sk, alice_pk, shared_secret)
```
```suggestion
        return (alice_sk, alice_pk)
```

vincentmacri replied:
Same thing in `bob_keygen`.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/key_exchange.py:
```diff
@@ -0,0 +1,56 @@
+    def bob_compute_shared_secret(self, bob_sk, alice_pk):
+        raise NotImplementedError
+
+    def alice_keygen(self):
```
```suggestion
    def alice_keygen(self) -> tuple[Any, Any]:
```

And add `from typing import Any`

vincentmacri commented on src/sage/crypto/public_key/key_exchange/key_exchange.py:
```diff
@@ -0,0 +1,56 @@
+        bob_pk = self.bob_public_key(bob_sk)
+        alice_shared_secret = self.alice_compute_shared_secret(alice_sk, bob_pk)
+        bob_shared_secret = self.bob_compute_shared_secret(bob_sk, alice_pk)
+        return (alice_sk, alice_pk, alice_shared_secret, bob_sk, bob_pk, bob_shared_secret)
```
```suggestion
    def do_key_exchange(self) -> tuple[Any, Any, Any, Any, Any]:
        alice_sk, alice_pk = self.alice_keygen()
        bob_sk, bob_pk = self.bob_keygen()
        alice_shared_secret = self.alice_compute_shared_secret(alice_sk, bob_pk)
        assert alice_shared_secret == self.bob_compute_shared_secret(bob_sk, alice_pk)
        return (alice_sk, alice_pk, bob_sk, bob_pk, alice_shared_secret)
```

vincentmacri replied:
No point in returning both shared secrets as the only way they aren't equal is if there is a bug.

Brian-Heckel replied:
My idea of making it return both shared_secrets is if someone implements `KeyExchange` the user can do a simple check if their key exchange works by running the function in the REPL and checking that the two shared secrets are equal. This would also make the testing code eaiser as we can just do
```python
for _ in range(tests):
    (_, _, alice_shared, _, _, bob_shared) = exchange.do_key_exchange()
    tester.assertEqual(alice_shared_secret, bob_shared_secret)
```
maybe I could reorder it as `alice_pk, alice_sk, bob_pk, bob_sk, alice_shared, bob_shared` so that it is eaiser to see on the user side.

vincentmacri replied:
Testing should be implemented by TestSuite anyway, and testing code should be very explicit as well. And computing the shared secret twice and doing an assert like I suggested still allows for easy testing. I think the main utility of this function is as a "give me everything" helper function.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/key_exchange.py:
```diff
@@ -0,0 +1,56 @@
+        bob_shared_secret = self.bob_compute_shared_secret(bob_sk, alice_pk)
+        return (alice_sk, alice_pk, alice_shared_secret, bob_sk, bob_pk, bob_shared_secret)
+       
+
```
```suggestion
    def _test_key_exchange(self, **options):
        tester = self._tester(**options)

        alice_sk, alice_pk = self.alice_keygen()                        
        bob_sk, bob_pk = self.bob_keygen()
        alice_shared_secret = self.alice_compute_shared_secret(alice_sk, bob_pk)
        bob_shared_secret = self.bob_compute_shared_secret(bob_sk, alice_pk)
        tester.assertEqual(alice_shared_secret, bob_shared_secret)
```
Run 5 key exchanges, error if any of them do not arrive at the same shared secret.

This will get picked up by `TestSuite`.

Brian-Heckel replied:
This code looks good, I'll try and implement a wrong key exchange and see the error diagnostic but I believe that the `assertEqual` will only show the result of the `alice_shared_secret` and `bob_shared_secret` values, I'll look into the `assertEqual` docs, but I would like to also show what the public and secret keys were that lead to a bug.

vincentmacri replied:
`assertEqual` has an optional `msg` keyword parameter, you can add the public and secret keys there

vincentmacri replied:
https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual

vincentmacri reviewed (changes requested):
Good start! Other than splitting anshel_anshel_goldfeld out into a separate PR, all suggestions are fairly minor.

vincentmacri commented on src/sage/crypto/public_key/key_exchange/key_exchange.py:
```diff
@@ -0,0 +1,56 @@
+from abc import ABC, abstractmethod
```
`from sage.crypto.public_key.key_exchange.key_exchange import KeyExchange` is awkward, but not really avoidable. `from sage.crypto.public_key.key_exchange.key_exchange_base import KeyExchangeBase` is slightly less awkward. Personally I prefer `abstract_key_exchange` but `_base` matches the rest of the Sage library better.

So rename this file and the class. Also change `CommutativeKeyExchange` to `CommutativeKeyExchangeBase`.

Brian-Heckel replied:
Resolved Through Commit 4c78dfc

Brian-Heckel committed:
WIP First Review Changes

- `alice_keygen` only returns public and secret key
- rename `compute_all_keys` all `do_key_exchange` and return only one shared secret key
- Type hints to `alice_keygen` and `do_key_exchange`

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

Brian-Heckel committed:
Removed anshel anshel goldfeld to Move to Different PR

Brian-Heckel committed:
First Modifications After Review

- Moved CommutativeKeyExchange to key_exchange.py
- FiniteFieldDH uses CommutativeKeyExchange
- Added the Base suffix to the classes to indicate to users that it
    is a base class
- Added _test_key_exchange() to KeyExchangeBase
- Removed the ABC inheritance from CommutativeKeyExchangeBase
- Added __dir__ method to CommutativeKeyExchangeBase for easier
    discoverability

Brian-Heckel committed:
Renamed key_exchange.py to key_exchange_base.py

Brian-Heckel committed:
Fixed Ruff Lint Errors

vincentmacri commented:
Looks good. I'll test the behaviour of `__dir__` locally, clean up for style, and maybe add a couple doctests later this week before merging.

vincentmacri committed:
Merge branch 'crypto' into brian-key-exchange
